src/Caching/Hybrid/src/Internal/DefaultHybridCache.CacheItem.cs

```diff
@@ -0,0 +1,45 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using Microsoft.Extensions.Caching.Memory;
+
+namespace Microsoft.Extensions.Caching.Hybrid.Internal;
+
+partial class DefaultHybridCache
+{
+    internal abstract class CacheItem
+    {
+        internal static readonly PostEvictionDelegate OnEviction = (key, value, reason, state) =>
+        {
+            if (value is CacheItem item)
+            {
+                // in reality we only set this up for mutable cache items, as a mechanism
+                // to recycle the buffers
+                item.Release();
+            }
+        };
+
+        public virtual void Release() { } // for recycling purposes
+
+        public virtual bool NeedsEvictionCallback => false; // do we need to call Release when evicted?
+
+        public abstract bool TryReserveBuffer(out BufferChunk buffer);
+    }
+
+    internal abstract class CacheItem<T> : CacheItem
+    {
+        public abstract bool TryGetValue(out T value);
+
+        public T GetValue()
+        {
+            if (!TryGetValue(out var value))
+            {
+                Throw();
+            }
+            return value;
+
+            static void Throw() => throw new ObjectDisposedException("The cache item has been recycled before the value was obtained");
```

Comment:
Correct, it lets the calling method be inlined into its caller. See the "Use static throw helpers" section here: https://reubenbond.github.io/posts/dotnet-perf-tuning. I am not certain if the advice still holds, given all of the advancements in the JIT in the last 5 years

Reply:
> I am not certain if the advice still holds

It still holds.
With JIT's PGO capabilities the cold pathes would be moved to a cold block of machine code, but it's not split into a separate method. Thus manually splitting these rare paths into a separate method makes sense, as the hot-/happy-path method gets smaller and is potentially inlined.

For non-throwing rare path methods they can be marked with non-inlining, to make sure that the JIT won't inline these methods by coincidence.
For throwing methods non-inlining shouldn't be applied, in order for the JIT to see that the method won't return (because of the `throw`). If at these methods non-inlining would be applied the JIT wouldn't examine the method at all, and just emit a call to it.

Reply:
The reason the JIT doesn't inline methods with throws is that it wants to keep them on the stack in the event of an exception?